rfcs/text/0004_application_service_mounting.md

```diff
@@ -0,0 +1,237 @@
+- Start Date: 2019-05-10
+- RFC PR: (leave this empty)
+- Kibana Issue: (leave this empty)
+
+# Summary
+
+A front-end service to manage registration and root-level routing for
+first-class applications.
+
+# Basic example
+
+```tsx
+import React from 'react';
+import ReactDOM from 'react-dom';
+
+import { MyApp } from './components';
```

Comment:
My only concern with the standard importing approach has to do with statefulness. It is possible to import something here that could break something on another page without ever mounting. We just have to watch out for this.

---

This also has the potential to increase the size of the upfront load if not done properly. We could recommend importing certain items dynamically:

```js
mount: async (context, targetDomElement) => {
  const { MyApp } = await import('./components');

  ReactDOM.render(
    <MyAppRoot mountContext={context} deps={pluginStart} />,
    targetDomElement
  );
```

But that just shifts loading around, so different tradeoff. Just things to think about.

Reply:
> My only concern with the standard importing approach has to do with statefulness. It is possible to import something here that could break something on another page without ever mounting. We just have to watch out for this.

Definitely possible, but I think that's true regardless. Technically, you could do something bad in any JS file we import (like your top-level `public/index.js` file). We can't prevent plugins from doing every bad thing possible.

> This also has the potential to increase the size of the upfront load if not done properly. We could recommend importing certain items dynamically

This is a good point and how I would like plugins load their UI code. I think it's worth updating this to reflect the ideal usage. I'll do that now.